Firestore/core/src/firebase/firestore/core/sync_engine.h

```diff
@@ -0,0 +1,293 @@
+     * applies the query filters and limits to determine the most correct
+     * possible results.
+     */
+    View& view() {
+      return view_;
+    }
+
+   private:
+    Query query_;
+    model::TargetId target_id_;
+    nanopb::ByteString resume_token_;
+    View view_;
+  };
+
+  /** Tracks a limbo resolution. */
+  class LimboResolution {
+   public:
+    LimboResolution() = default;
+
+    explicit LimboResolution(const model::DocumentKey& key) : key{key} {
+    }
+
+    model::DocumentKey key;
+
+    /**
+     * Set to true once we've received a document. This is used in
+     * remoteKeysForTarget and ultimately used by `WatchChangeAggregator` to
+     * decide whether it needs to manufacture a delete event for the target once
+     * the target is CURRENT.
+     */
+    bool document_received = false;
+  };
+
+  void AssertCallbackExists(absl::string_view source);
+
+  ViewSnapshot InitializeViewAndComputeSnapshot(
+      const local::QueryData& query_data);
+
+  void RemoveAndCleanupQuery(const std::shared_ptr<QueryView>& query_view);
+
+  void RemoveLimboTarget(const model::DocumentKey& key);
+
+  void EmitNewSnapshotsAndNotifyLocalStore(
+      const model::MaybeDocumentMap& changes,
+      const absl::optional<remote::RemoteEvent>& maybe_remote_event);
+
+  /** Updates the limbo document state for the given target_id. */
+  void UpdateTrackedLimboDocuments(
+      const std::vector<LimboDocumentChange>& limbo_changes,
+      model::TargetId target_id);
+
+  void TrackLimboChange(const LimboDocumentChange& limbo_change);
+
+  void NotifyUser(model::BatchId batch_id, util::Status status);
+
+  /**
+   * Triggers callbacks waiting for this batch id to get acknowledged by
+   * server, if there are any.
+   */
+  void TriggerPendingWriteCallbacks(model::BatchId batch_id);
+  void FailOutstandingPendingWriteCallbacks(absl::string_view message);
+
+  /** The local store, used to persist mutations and cached documents. */
+  FSTLocalStore* local_store_;
+
+  /** The remote store for sending writes, watches, etc. to the backend. */
+  remote::RemoteStore* remote_store_ = nullptr;
+
+  auth::User current_user_;
+  SyncEngineCallback* sync_engine_callback_ = nullptr;
+
+  /**
+   * Used for creating the TargetId for the listens used to resolve limbo
+   * documents.
+   */
+  TargetIdGenerator target_id_generator_;
+
+  /** Stores user completion blocks, indexed by User and BatchId. */
+  std::unordered_map<auth::User,
+                     std::unordered_map<model::BatchId, util::StatusCallback>,
+                     auth::HashUser>
+      mutation_callbacks_;
+
+  /** Stores user callbacks waiting for pending writes to be acknowledged. */
+  std::unordered_map<model::BatchId, std::vector<util::StatusCallback>>
+      pending_writes_callbacks_;
+
+  /** QueryViews for all active queries, indexed by query. */
+  std::unordered_map<Query, std::shared_ptr<QueryView>> query_views_by_query_;
```

Comment:
I think this can be `std::shared_ptr<const QueryView` (here and below, also in the `std::make_shared`).

Note that `const` is particularly valuable with shared pointers -- a shared mutable object is hard to reason about, while a shared constant is pretty harmless.

Reply:
Right, I missed that the `View` has to be non-`const`.